### PR TITLE
camel service registry

### PR DESCRIPTION
**Pull request by lburgazzoli**
This PR is the initial work to ease the process to expose routes as services [CAMEL-10793](https://issues.apache.org/jira/browse/CAMEL-10793).

A simple example is:

```java
from("service:my-service:undertow:http://0.0.0.0:8080/the/context/path")
    .log("Route ${routeId} has been invoked");
```

As undertow component, supports service registration, this endpoint will be registered to a given Service Registry (we now have a binding for consul, zookeeper and spring-cloud) as `my-service` along with some metadata such as the protocol and the context path. This allow to invoke the service on another jvm instance by name, like:

```java
from("direct:call")
    .serviceCall("my-service")
```

**NOTE**: The implementation is not yet complete so you still need to do some manual configuration and the number of supported component that support service registration is limited to those extending `camel-http-common` and `undertow`  .
**NOTE**: integration with spring-cloud is not as easy as it should be so I had to introduce two new modules: camel-spring-cloud-zookeeper and camel-spring-cloud-consul.

Any feedback would be really appreciated.

### CODE REVIEW

File: camel-core/src/main/java/org/apache/camel/impl/cloud/ServiceRegistrationRoutePolicy.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+import org.apache.camel.api.management.ManagedResource;
+import org.apache.camel.cloud.DiscoverableService;
+import org.apache.camel.cloud.ServiceDefinition;
+import org.apache.camel.cloud.ServiceRegistry;
+import org.apache.camel.support.RoutePolicySupport;
+import org.apache.camel.util.ObjectHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ManagedResource(description = "Clustered Route policy using")
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/impl/cloud/ServiceRegistrationRoutePolicy.java]: Update description

File: camel-core/src/main/java/org/apache/camel/impl/cloud/ServiceDefinitionHelper.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.impl.cloud;
+
+public final class ServiceDefinitionHelper {
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/impl/cloud/ServiceDefinitionHelper.java]: Seems like not in use

File: camel-core/src/test/java/org/apache/camel/impl/cloud/ServiceRegistryTest.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.camel.ContextTestSupport;
+import org.apache.camel.cloud.ServiceDefinition;
+import org.apache.camel.model.RouteDefinition;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ServiceRegistryTest  extends ContextTestSupport {
```
Comment (davsclaus) [inline camel-core/src/test/java/org/apache/camel/impl/cloud/ServiceRegistryTest.java]: Double space

File: components/camel-service/src/main/docs/service-component.adoc [hunk 4]
```diff
@@ -0,0 +1,85 @@
+[[service-component]]
+== Service Component
+== Master Component
```
Comment (davsclaus) [inline components/camel-service/src/main/docs/service-component.adoc]: Copy paste issues here, and duplicate lines

File: components/camel-service/src/main/java/org/apache/camel/component/service/ServiceEndpoint.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        this.delegateEndpoint = getCamelContext().getEndpoint(delegateUri);
+
+        // The service properties set on uri override parameter provided by a
+        // an endpoint of type DiscoverableService.
+        this.serviceDefinition = computeServiceDefinition(component.getCamelContext(), delegateEndpoint);
+    }
+
+    @ManagedAttribute(description = "The consumer endpoint to expose as a service", mask = true)
+    @Override
+    public Endpoint getEndpoint() {
```
Comment (davsclaus) [inline components/camel-service/src/main/java/org/apache/camel/component/service/ServiceEndpoint.java]: I dont think complex objects over JMX is good, eg Endpoint. Should be the string uri of the endpoint returned, eg String type

File: components/camel-service/src/main/java/org/apache/camel/component/service/ServiceEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        return this.delegateEndpoint;
+    }
+
+    @ManagedAttribute(description = "The service definition", mask = true)
+    public ServiceDefinition getServiceDefinition() {
```
Comment (davsclaus) [inline components/camel-service/src/main/java/org/apache/camel/component/service/ServiceEndpoint.java]: See previous comment.

**Comment by davsclaus** [comment]
Great start, a lot of great stuff in there - keep it going

**Comment by lburgazzoli** [comment]
merged manually

**Pull request closed without merging**